### ISSUE TITLE
corrupt JPEG data: # extraneous bytes before marker 0x##

### ISSUE THREAD

**Issue by pjsanfil** [opencv/opencv#9477]
##### System information (version)
- OpenCV => 3.3
- Operating System / Platform => Ubuntu 16.04.3 LTS (x86_64)
- Compiler => gcc 5.4.0 20160609

##### Detailed description
When accessing a Logitech C310 webcam with Java OpenCV 3.3 the following error message is repeatedly printed to the terminal (with varying values for number of bytes and marker)
`Corrupt JPEG data: 2 extraneous bytes before marker 0xd3`

The error message comes from 3rdparty/libjpeg/jdmarker.c
usage of JWRN_EXTRANEOUS_DATA

The problem appears to be caused by a bug with the libjpeg that ships with OpenCV. A fix is to build OpenCV with the option `WITH_JPEG` switched OFF causing OpenCV to use the platform's copy of libjpeg.

I recommend updating the instructions [here](http://docs.opencv.org/2.4/doc/tutorials/introduction/desktop_java/java_dev_intro.html)
to include `WITH_JPEG=OFF`

##### Steps to reproduce
Build OpenCV for Java and run in Linux using the settings recommended in the [Introduction to Java Development](http://docs.opencv.org/2.4/doc/tutorials/introduction/desktop_java/java_dev_intro.html)

**Comment by alalek** [comment]
> bug with the libjpeg that ships with OpenCV
> 3rdparty/libjpeg

Small clarification: this warning is reproduced with system libjpeg libraries too.

I'm observed that these warnings are not showed for each frame. So It is hard to say what is source of problem.
Anyway result is not corrupted.

Could you check this workaround on your hardware: #9479 ?

**Comment by pjsanfil** [comment]
I verified the workaround does fix this issue on my hardware. Thank you.

**Comment by pjsanfil** [comment]
I think this needs to be reopened because the issue still occurs once the camera is switched into MJPEG codec. The workaround  causes it to default to YUYV codec but once switched to MJPEG the problem still occurs.

**Comment by anandcu3** [comment]
Any update on this issue? This is occurring on Python 3.5.2, Opencv 3.3.0,  Ubuntu 16.04.4 LTS (GNU/Linux 3.10.105-141 armv7l).

Anyway to at least hide these warnings apart from switching to YUYV codec.

**Comment by x92127** [comment]
I am also experiencing the same issue.  
Python 2.7.6
OpenCV 2.4.13
Ubuntu 14.04.1 LTS
Is anyone actively working on this?

**Comment by fzcooper** [comment]
This is a problem with libjpeg, not opencv. Line 1040 in opencv/3rdparty/libjpeg/jdmarker.c `WARNMS2(cinfo, JWRN_EXTRANEOUS_DATA, cinfo->marker->discarded_bytes, c);` can be deleted or commented out. Recompile and your nuisance messages should be eliminated.

Maybe the USB webcam doesn't properly implement mjpeg, is sending comments, or some other issue. I haven't had any problems due to the extraneous data and discovering the root of the problem would be pretty involved, so I am happy to mute that warning since the extraneous bytes have caused zero functional issues so far. Periodically a frame will be corrupt enough it causes visual artifacts in the rendering but it is infrequent and minor enough it doesn't affect my application.

**Comment by alalek** [comment]
> Line 1040 in opencv/3rdparty/libjpeg/jdmarker.c

OpenCV prefers using of system libjpeg by default (instead of builtin from `3rdparty`). libjpeg from system (Fedora 26 at least) produces the same warnings too. We can't patch system libjpeg for muting of these warnings.


> Maybe the USB webcam doesn't properly implement mjpeg

Maybe.


> Periodically a frame will be corrupt enough

I didn't observe significant frames corruption. Could you provide some details?


> it causes visual artifacts

IMHO, console messages have less impact than that.

**Comment by fzcooper** [comment]
> We can't patch system libjpeg for muting of these warnings.

Compiling with the BUILD_JPEG option will cause opencv to build and use its own version. My build script uses sed to find and delete the line mentioned in my previous post and then compile the modified libjpeg, disabling the extraneous bytes warnings.

> I didn't observe significant frames corruption. Could you provide some details?

It's very minor. From hours and hours of video I will occasionally save an image that is tinted purple or has funny shapes or colors embedded in it.

> IMHO, console messages have less impact than that.

Yes, depending on how your software is used. My system is an embedded platform miles from any roads or civilization and is rarely interacted with except via console. The messages cause too much interference when monitoring the parameters presented via the console.

**Comment by SanthoshRajendiran** [comment]
In my use case, all frames are essentially needed. Just to make sure, in this case, if the warning is noticed, will there be any data loss, simple question, will that be considered as an image read?

**Comment by simontamayo** [comment]
I had the same problem while running cv2.VideoCapture(0) and I solved it by re-installing openCv: 
`pip install opencv-python`

**Comment by AjinkyaPrabhu** [comment]
> I had the same problem while running cv2.VideoCapture(0) and I solved it by re-installing openCv:
> `pip install opencv-python`

yeah this works

**Comment by Fazankabir** [comment]
> > I had the same problem while running cv2.VideoCapture(0) and I solved it by re-installing openCv:
> > `pip install opencv-python`
> 
> yeah this works

thank you @simontamayo  @AjinkyaPrabhu did you guys try this on linux?

**Comment by simontamayo** [comment]
> > > I had the same problem while running cv2.VideoCapture(0) and I solved it by re-installing openCv:
> > > `pip install opencv-python`
> > 
> > 
> > yeah this works
> 
> thank you @simontamayo @AjinkyaPrabhu did you guys try this on linux?

@Fazankabir : yes, I did it on my Ubuntu 18.04.1
Cheers,

**Comment by Fazankabir** [comment]
> > > > I had the same problem while running cv2.VideoCapture(0) and I solved it by re-installing openCv:
> > > > `pip install opencv-python`
> > > 
> > > 
> > > yeah this works
> > 
> > 
> > thank you @simontamayo @AjinkyaPrabhu did you guys try this on linux?
> 
> @Fazankabir : yes, I did it on my Ubuntu 18.04.1
> Cheers,

@simontamayo I have compiled the OpenCV on raspberry pi following  steps from here https://www.pyimagesearch.com/2017/09/04/raspbian-stretch-install-opencv-3-python-on-your-raspberry-pi/  and now when i have connected the USB logitech camera I am getting the same issue. any help? Thanks in advance

**Comment by ab791520b4** [comment]
> This is a problem with libjpeg, not opencv. Line 1040 in opencv/3rdparty/libjpeg/jdmarker.c `WARNMS2(cinfo, JWRN_EXTRANEOUS_DATA, cinfo->marker->discarded_bytes, c);` can be deleted or commented out. Recompile and your nuisance messages should be eliminated.

I can confirm this. Though for opencv-python I needed to patch `opencv/3rdparty/libjpeg-turbo/jdmarker.c` instead.

**Comment by alalek** [comment]
> opencv/3rdparty

Problem is reproducible even with system libjpeg binaries so patching of OpenCV's fallback code for libjpeg can't solve this issue.

**Comment by en3sis** [comment]
What worked for me while using Python 3.7.3 on the Raspberry PI while using an USB Logitech Webcam was to install the following dependencies:

`pip3 install opencv-contrib-python; sudo apt-get install -y libatlas-base-dev libhdf5-dev libhdf5-serial-dev libatlas-base-dev libjasper-dev  libqtgui4  libqt4-test`

Now the warning is gone.

**Comment by KavitShah1998** [comment]
**For those who are using a web-cam to capture images/video feed.**  

This issue occurs with a few of _Logitech webcams_ as seen in [Issue2194](https://github.com/tensorflow/models/issues/2194#issuecomment-335216819) . 

I faced the same problem with Logitech C270 web-cam. I tried running the same code and replacing the Logitech C270p web-cam with Nulaxy C900 & it ran without any of those warnings.

The reason could more likely be something highlighted by @fzcooper [here](https://github.com/opencv/opencv/issues/9477#issuecomment-390679150)


Best,
Kavit Shah

**Comment by aleksandaratanasov** [comment]
I can provide samples to reproduce the error without any webcam. Just download the [Kaggle Cats and Dogs dataset](https://www.microsoft.com/en-us/download/details.aspx?id=54765) and iterate over all images, loading each with OpenCV (I used the Python API). The following code snippet (I skipped the imports)

```
pets = ['PetImages/Cats', 'PetImages/Dogs']
for pet_dir in pets:
    for img_path in os.listdir(pet_dir):
          pet_img_path = os.path.join(pet_path, img_path)
          try:    
              pet_img = cv2.imread(pet_img_path, cv2.IMREAD_GRAYSCALE)
              pet_img = cv2.resize(pet_img, (self.img_size, self.img_size), interpolation=cv2.INTER_CUBIC)
          except Exception as ex:
              print('Failed to load', pet_img_path)
```

yields

```
Loading pet data from PetImages/Cat
Failed to load PetImages/Cat\10125.jpg
Failed to load PetImages/Cat\10404.jpg
Failed to load PetImages/Cat\10501.jpg
Failed to load PetImages/Cat\10820.jpg
Failed to load PetImages/Cat\11210.jpg
Failed to load PetImages/Cat\11565.jpg
Failed to load PetImages/Cat\11874.jpg
Failed to load PetImages/Cat\11935.jpg
Corrupt JPEG data: 242 extraneous bytes before marker 0xd9
Failed to load PetImages/Cat\140.jpg
Corrupt JPEG data: 217 extraneous bytes before marker 0xd9
Corrupt JPEG data: 133 extraneous bytes before marker 0xd9
Failed to load PetImages/Cat\2663.jpg
Failed to load PetImages/Cat\3300.jpg
Failed to load PetImages/Cat\3491.jpg
Failed to load PetImages/Cat\4833.jpg
Failed to load PetImages/Cat\5553.jpg
Corrupt JPEG data: 104 extraneous bytes before marker 0xd9
Failed to load PetImages/Cat\660.jpg
Failed to load PetImages/Cat\666.jpg
Corrupt JPEG data: 1158 extraneous bytes before marker 0xd9
Failed to load PetImages/Cat\7968.jpg
Failed to load PetImages/Cat\7978.jpg
Failed to load PetImages/Cat\8470.jpg
Failed to load PetImages/Cat\850.jpg
Failed to load PetImages/Cat\9171.jpg
Failed to load PetImages/Cat\936.jpg
Failed to load PetImages/Cat\9565.jpg
Failed to load PetImages/Cat\9778.jpg
Failed to load PetImages/Cat\Thumbs.db
Loading pet data from PetImages/Dog
Failed to load PetImages/Dog\10158.jpg
Failed to load PetImages/Dog\10401.jpg
Failed to load PetImages/Dog\10747.jpg
Failed to load PetImages/Dog\10797.jpg
Failed to load PetImages/Dog\11410.jpg
Failed to load PetImages/Dog\11675.jpg
Failed to load PetImages/Dog\11702.jpg
Failed to load PetImages/Dog\11849.jpg
Failed to load PetImages/Dog\11853.jpg
Corrupt JPEG data: 402 extraneous bytes before marker 0xd9
Failed to load PetImages/Dog\1308.jpg
Failed to load PetImages/Dog\1866.jpg
Failed to load PetImages/Dog\2384.jpg
Failed to load PetImages/Dog\2688.jpg
Failed to load PetImages/Dog\2877.jpg
Corrupt JPEG data: 232 extraneous bytes before marker 0xd9
Failed to load PetImages/Dog\3136.jpg
Failed to load PetImages/Dog\3288.jpg
Failed to load PetImages/Dog\3588.jpg
Corrupt JPEG data: 1 extraneous bytes before marker 0xd9
Corrupt JPEG data: 164 extraneous bytes before marker 0xd9
Warning: unknown JFIF revision number 0.00
Corrupt JPEG data: 1408 extraneous bytes before marker 0xd9
Failed to load PetImages/Dog\4367.jpg
Failed to load PetImages/Dog\5604.jpg
Failed to load PetImages/Dog\5736.jpg
Corrupt JPEG data: 259 extraneous bytes before marker 0xd9
Failed to load PetImages/Dog\6059.jpg
Failed to load PetImages/Dog\6238.jpg
Failed to load PetImages/Dog\6718.jpg
Corrupt JPEG data: 2232 extraneous bytes before marker 0xd9
Failed to load PetImages/Dog\7112.jpg
Failed to load PetImages/Dog\7133.jpg
Failed to load PetImages/Dog\7369.jpg
Failed to load PetImages/Dog\7459.jpg
Corrupt JPEG data: 70 extraneous bytes before marker 0xd9
Failed to load PetImages/Dog\7969.jpg
Failed to load PetImages/Dog\8730.jpg
Failed to load PetImages/Dog\9188.jpg
```

for all images in the dataset.

**Comment by nicolas-abril-art** [comment]
> 

Except for Cats/666.jpg which is an empty file, all the others that failed to load seem to be GIF files with the incorrect extension.

**Comment by halqadasi** [comment]
We read each JPEG image in the specified folder using OpenCV and overwrite it by re-saving with a quality setting of 95. This re-encoding process fixes any corruption issues by standardizing the image data format.

```
    for filename in tqdm(image_files, desc="Re-encoding images"):

        image_path = os.path.join(folder_path, filename)
        img = cv2.imread(image_path)

        if img is not None:
           # Re-encode and overwrite the image
           cv2.imwrite(image_path, img, [int(cv2.IMWRITE_JPEG_QUALITY), 95])
        else:
            print(f"Failed to read image data: {image_path}")
```

**This solution fixed my problem after long time of debugging.**
